src/zenml/integrations/tensorflow/visualizers/tensorboard_visualizer.py

```diff
@@ -0,0 +1,216 @@
+                    # Daemon service functionality is currently not supported on Windows
+                    print(
+                        "You can run:\n"
+                        f"[italic green]    tensorboard --logdir {logdir}"
+                        "[/italic green]\n"
+                        "...to visualize the Tensorboard logs for your trained model."
+                    )
+                else:
+                    # start a new Tensorboard server
+                    service = TensorboardService(
+                        TensorboardServiceConfig(
+                            logdir=logdir,
+                        )
+                    )
+                    service.start(timeout=20)
+                    if service.endpoint.status.port:
+                        self.visualize_tensorboard(
+                            service.endpoint.status.port, height
+                        )
+                return
+
+    def visualize_tensorboard(
+        self,
+        port: int,
+        height: int,
+    ) -> None:
+        """Generate a visualization of a Tensorboard.
+
+        Args:
+            port: the TCP port where the Tensorboard server is listening for
+                requests.
+            height: Height of the generated visualization.
+            logdir: The logdir location for the Tensorboard server.
+        """
+        if Environment.in_notebook():
+
+            notebook.display(port, height=height)
+            return
+
+        print(
+            "You can visit:\n"
+            f"[italic green]    http://localhost:{port}/[/italic green]\n"
+            "...to visualize the Tensorboard logs for your trained model."
+        )
+
+    def stop(
+        self,
+        object: StepView,
+    ) -> None:
+        """Stop the Tensorboard server previously started for a pipeline step.
+
+        Args:
+            object: StepView fetched from run.get_step().
+        """
+        for _, artifact_view in object.outputs.items():
+            # filter out anything but model artifacts
+            if artifact_view.type == ModelArtifact.TYPE_NAME:
+                logdir = os.path.dirname(artifact_view.uri)
+
+                # first check if a Tensorboard server is already running for
+                # the same logdir location and use that one
+                running_server = self.find_running_tensorboard_server(logdir)
+                if not running_server:
+                    return
+
+                logger.debug(
+                    "Stopping tensorboard server with PID '%d' ...",
+                    running_server.pid,
+                )
+                try:
+                    p = psutil.Process(running_server.pid)
+                except psutil.Error:
+                    logger.error(
+                        "Could not find process for PID '%d' ...",
+                        running_server.pid,
+                    )
+                    continue
+                p.kill()
```

Comment:
Just for clarification: We need to implement it like this right now because ZenML doesn't track running services and the service isn't stored as an artifact somewhere right? Otherwise we could use `tensorboard_service.stop()` instead of killing the process manually?

Reply:
Exactly ! Implementing extensible service registry and tracking is on my TODO list for the service abstraction. When that is done, ZenML should be able to load all running services on startup and this code will be replaced with a simple service registry lookup followed by a `service.stop()`.